### PR TITLE
Add address setting

### PR DESCRIPTION
**Pull request by cyberliem**
Add address setting to reorganize core setting

### CODE REVIEW

**Comment by cyberliem** [comment]
[Add address setting to setting package.](https://trello.com/c/fL9ivLQu/59-add-address-setting-to-setting-package)

File: blockchain/blockchain.go [hunk 1]
```diff
@@ -114,15 +106,35 @@ func (self *Blockchain) GetAddresses() (*common.Addresses, error) {
...
+		return nil, err
+	}
+	burnerAddr, err := self.setting.GetAddress("burner")
+	if err != nil {
+		return nil, err
+	}
+	networkAddr, err := self.setting.GetAddress("network")
+	if err != nil {
+		return nil, err
+	}
```
Comment (favadi) [inline blockchain/blockchain.go]: @cyberliem please make those ""reserve"/"burner"/"network"... into constants, they are used everywhere.

File: blockchain/blockchain.go [hunk 2]
```diff
@@ -652,29 +695,48 @@ func (self *Blockchain) GetPricingMethod(inputData string) (*abi.Method, error)
...
 
-func NewBlockchain(
-	base *blockchain.BaseBlockchain,
-	wrapperAddr, pricingAddr, burnerAddr,
-	networkAddr, reserveAddr, whitelistAddr ethereum.Address,
-	setting Setting) (*Blockchain, error) {
+func NewBlockchain(base *blockchain.BaseBlockchain, setting Setting) (*Blockchain, error) {
+	wrapperAddr, err := setting.GetAddress("wrapper")
+	if err != nil {
+		log.Panicf("Can not get wrapper Address %s", err)
```
Comment (favadi) [inline blockchain/blockchain.go]: the error should be returned to caller, it will decide if it panic or not.

File: blockchain/blockchain.go [hunk 3]
```diff
@@ -652,29 +695,48 @@ func (self *Blockchain) GetPricingMethod(inputData string) (*abi.Method, error)
...
+	}
 	log.Printf("wrapper address: %s", wrapperAddr.Hex())
 	wrapper := blockchain.NewContract(
 		wrapperAddr,
 		filepath.Join(common.CurrentDir(), "wrapper.abi"),
 	)
+	reserveAddr, err := setting.GetAddress("reserve")
+	if err != nil {
+		log.Panicf("Can not get reserve Address %s", err)
```
Comment (favadi) [inline blockchain/blockchain.go]: the error should be returned to caller, it will decide if it panic or not.

File: blockchain/blockchain.go [hunk 4]
```diff
@@ -652,29 +695,48 @@ func (self *Blockchain) GetPricingMethod(inputData string) (*abi.Method, error)
...
+	}
 	log.Printf("reserve address: %s", reserveAddr.Hex())
 	reserve := blockchain.NewContract(
 		reserveAddr,
 		filepath.Join(common.CurrentDir(), "reserve.abi"),
 	)
+	pricingAddr, err := setting.GetAddress("pricing")
+	if err != nil {
+		log.Panicf("Can not get pricing Address %s", err)
```
Comment (favadi) [inline blockchain/blockchain.go]: the error should be returned to caller, it will decide if it panic or not.

File: blockchain/blockchain.go [hunk 5]
```diff
@@ -652,29 +695,48 @@ func (self *Blockchain) GetPricingMethod(inputData string) (*abi.Method, error)
...
+	}
 	log.Printf("pricing address: %s", pricingAddr.Hex())
 	pricing := blockchain.NewContract(
 		pricingAddr,
 		filepath.Join(common.CurrentDir(), "pricing.abi"),
 	)
-
+	burnerAddr, err := setting.GetAddress("burner")
+	if err != nil {
+		log.Panicf("Can not get burner address %s", err)
```
Comment (favadi) [inline blockchain/blockchain.go]: the error should be returned to caller, it will decide if it panic or not.

File: blockchain/blockchain.go [hunk 6]
```diff
@@ -652,29 +695,48 @@ func (self *Blockchain) GetPricingMethod(inputData string) (*abi.Method, error)
...
+	}
 	log.Printf("burner address: %s", burnerAddr.Hex())
+	networkAddr, err := setting.GetAddress("network")
+	if err != nil {
+		log.Panicf("Can not get network address %s", err)
```
Comment (favadi) [inline blockchain/blockchain.go]: the error should be returned to caller, it will decide if it panic or not.

File: blockchain/blockchain.go [hunk 7]
```diff
@@ -652,29 +695,48 @@ func (self *Blockchain) GetPricingMethod(inputData string) (*abi.Method, error)
...
+	}
 	log.Printf("network address: %s", networkAddr.Hex())
+	whitelistAddr, err := setting.GetAddress("whitelist")
+	if err != nil {
+		log.Panicf("Can not get whitelist address %s", err)
```
Comment (favadi) [inline blockchain/blockchain.go]: the error should be returned to caller, it will decide if it panic or not.

File: blockchain/blockchain.go (hunk 1 above)
Comment (favadi) [reply blockchain/blockchain.go]: Actually as I reviewed, those address type should be defined as enums like that.

```golang
type AddressType int
const (
    Reserve = iota
    Burner
    Network
    ....
)

const addressTypes = ["reserve", "bundle", ...]

func (at addressType) String() {
    return addressTypes[at]
}

func IsAddressTypeName(addr string) {
    // check if addr is valid type name
}
```

This way we can use compiler type check, instead of passing string around, which is error prone.

For example in golang builtin https://github.com/golang/go/blob/3b137dd2df19c261a007b8a620a2182cd679d700/src/cmd/internal/objabi/head.go

File: cmd/configuration/getconfig.go [hunk 8]
```diff
@@ -135,9 +131,9 @@ func GetConfig(kyberENV string, authEnbl bool, endpointOW string, noCore, enable
 	}
 
 	if enableStat {
-		config.AddStatConfig(setPath, addressConfig)
+		config.AddStatConfig(setPath)
 	}
-
+	//TODO : remove addressconfig, add exchange to setting
```
Comment (favadi) [inline cmd/configuration/getconfig.go]: This is reminder for doing in other task? If yes, please add the link to this line in that card.

File: core/reserve_core.go [hunk 9]
```diff
@@ -205,7 +205,10 @@ func (self ReserveCore) Withdraw(
 	} else {
 		err = sanityCheckAmount(exchange, token, amount)
 		if err == nil {
-			id, err = exchange.Withdraw(token, amount, self.rm, timepoint)
+			reserveAddr, err := self.setting.GetAddress("reserve")
```
Comment (favadi) [inline core/reserve_core.go]: It is not really related to the task, but could you changed the codes here to handle the odd case first (if there is error, just return it before returning to normal path). It is kind of hard to follow the logic here.

File: core/reserve_core_test.go [hunk 10]
```diff
@@ -139,10 +143,16 @@ func (self testActivityStorage) HasPendingDeposit(token common.Token, exchange c
 }
 
 func getTestCore(hasPendingDeposit bool) *ReserveCore {
+	tmpDir, err := ioutil.TempDir("", "test_pwi_equation_v2")
```
Comment (favadi) [inline core/reserve_core_test.go]: "test_pwi_equation_v2" does not make sense in this test. Bad copy/pasta?

File: data/fetcher/fetcher_setting.go [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+)
+
+type Setting interface {
+	//	GetInternalTokens() ([]common.Token, error)
+	//	GetActiveTokens() ([]common.Token, error)
+	//	GetTokenByAddress(addr ethereum.Address) (common.Token, error)
+	//	ETHToken() common.Token
+	//	GetActiveTokenByID(id string) (common.Token, error)
+	GetAddress(name string) (ethereum.Address, error)
+	//	GetAddresses(setName string) ([]ethereum.Address, error)
```
Comment (favadi) [inline data/fetcher/fetcher_setting.go]: why those commented lines?

File: http/settings_handler.go [hunk 12]
```diff
@@ -5,6 +5,8 @@ import (
 	"fmt"
 	"strconv"
 
+	ethereum "github.com/ethereum/go-ethereum/common"
+
```
Comment (favadi) [inline http/settings_handler.go]: empty line.

File: http/settings_handler.go [hunk 13]
```diff
@@ -99,3 +101,33 @@ func (self *HTTPServer) TokenSettings(c *gin.Context) {
...
+func (self *HTTPServer) UpdateAddress(c *gin.Context) {
+	postForm, ok := self.Authenticated(c, []string{"name", "address"}, []Permission{RebalancePermission, ConfigurePermission})
+	if !ok {
+		return
+	}
+	addrStr := postForm.Get("address")
+	name := postForm.Get("name")
+	addr := ethereum.HexToAddress(addrStr)
+	err := self.setting.UpdateAdress(name, addr)
+	if err != nil {
```
Comment (favadi) [inline http/settings_handler.go]: please use `if err := xxx; err != nil` form.

File: http/settings_handler.go [hunk 14]
```diff
@@ -99,3 +101,33 @@ func (self *HTTPServer) TokenSettings(c *gin.Context) {
...
+func (self *HTTPServer) AddAddressToSet(c *gin.Context) {
+	postForm, ok := self.Authenticated(c, []string{"setname", "address"}, []Permission{RebalancePermission, ConfigurePermission})
+	if !ok {
+		return
+	}
+	addrStr := postForm.Get("address")
+	setName := postForm.Get("setname")
+	addr := ethereum.HexToAddress(addrStr)
+	err := self.setting.AddAddressToSet(setName, addr)
+	if err != nil {
```
Comment (favadi) [inline http/settings_handler.go]: please use `if err := xxx; err != nil` form.

File: settings/address_APIs.go [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+import (
+	ethereum "github.com/ethereum/go-ethereum/common"
+)
+
+func (setting *Settings) UpdateAdress(name string, address ethereum.Address) error {
+	return setting.Address.Storage.UpdateOneAddress(name, address.Hex())
+}
+
+func (setting *Settings) GetAddress(name string) (ethereum.Address, error) {
+	addr, err := setting.Address.Storage.GetAddress(name)
```
Comment (favadi) [inline settings/address_APIs.go]: we should return err right away here?

File: settings/address_setting.go [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"log"
+
+	settingstorage "github.com/KyberNetwork/reserve-data/settings/storage"
+)
+
+type AddressConfig struct {
```
Comment (favadi) [inline settings/address_setting.go]: from now on please add comment docs.

File: settings/address_setting.go [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+	Network            string   `json:"network"`
+	Wrapper            string   `json:"wrapper"`
+	Pricing            string   `json:"pricing"`
+	FeeBurner          string   `json:"feeburner"`
+	Whitelist          string   `json:"whitelist"`
+	ThirdPartyReserves []string `json:"third_party_reserves"`
+	SetRate            string   `json:"setrate"`
+}
+
+type AddressSetting struct {
```
Comment (favadi) [inline settings/address_setting.go]: from now on please add comment docs.

File: settings/address_setting.go [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+	Storage AddressStorage
+}
+
+func NewAddressSetting(dbPath string) *AddressSetting {
```
Comment (favadi) [inline settings/address_setting.go]: from now on please add comment docs.

File: settings/address_setting.go [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+	BoltAddressStorage, err := settingstorage.NewBoltAddressStorage(dbPath)
+	if err != nil {
+		log.Panicf("Setting Init: Can not create bolt address storage (%s)", err)
```
Comment (favadi) [inline settings/address_setting.go]: this function should return error instead of panic.

File: settings/storage/address_bolt.go [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+type BoltAddressStorage struct {
+	db *bolt.DB
+}
+
+func NewBoltAddressStorage(dbPath string) (*BoltAddressStorage, error) {
+	var err error
+	var db *bolt.DB
+	db, err = bolt.Open(dbPath, 0600, nil)
+	if err != nil {
+		panic(err)
```
Comment (favadi) [inline settings/storage/address_bolt.go]: please return the error.

File: settings/storage/address_bolt.go [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+		if _, uErr := tx.CreateBucketIfNotExists([]byte(ADDRESS_SETTING_BUCKET)); err != nil {
+			return uErr
+		}
+		if _, uErr := tx.CreateBucketIfNotExists([]byte(ADDRESS_SET_SETTING_BUCKET)); err != nil {
+			return uErr
+		}
+		return nil
+	})
+	if err != nil {
+		panic(err)
```
Comment (favadi) [inline settings/storage/address_bolt.go]: please return the error.

File: stat/fetcher.go [hunk 22]
```diff
@@ -149,7 +140,11 @@ func (self *Fetcher) FetchTxs(client http.Client) error {
 	if toBlock == 0 {
 		return errors.New("Can't get latest block nummber")
 	}
-	api := fmt.Sprintf("http://api.etherscan.io/api?module=account&action=txlist&address=%s&startblock=%d&endblock=%d&apikey=%s", self.pricingAddress.String(), fromBlock, toBlock, self.apiKey)
+	pricingAddress, err := self.setting.GetAddress("pricing")
+	if err != nil {
+		return err
+	}
+	api := fmt.Sprintf("http://api.etherscan.io/api?module=account&action=txlist&address=%s&startblock=%d&endblock=%d&apikey=%s", pricingAddress.String(), fromBlock, toBlock, self.apiKey)
```
Comment (favadi) [inline stat/fetcher.go]: why is that hard code api URL?

Comment (favadi) [reply stat/fetcher.go]: Looks like not related to your task.
@cyberliem is this a problem? If yes, please open a task for it.

File: core/reserve_core.go (hunk 9 above)
Comment (favadi) [reply core/reserve_core.go]: We will be notified once we enable golint checker.

File: settings/address_setting.go [hunk 23]
```diff
@@ -0,0 +1,93 @@
...
+	SetRate            string   `json:"setrate"`
+}
+
+// AddressSetting type defines component to handle all address setting in core.
+// It contains the storage interface used to query addresses.
+type AddressSetting struct {
+	Storage AddressStorage
+}
+
+func NewAddressSetting(dbPath string) (*AddressSetting, error) {
```
Comment (favadi) [inline settings/address_setting.go]: Now I notice this constructor is using a concrete implementation of AddressStorage interface instead of accepting the interface.

File: settings/address_storage.go [hunk 24]
```diff
@@ -0,0 +1,25 @@
+package settings
+
+//AddressStorage abstracts the storage of addresses.
+//Any implementation of AddressStorage must fulfilled these funcs
+//All params should be in lower case
+type AddressStorage interface {
+	//UpdateOneAddress stores an address, with its functional name(network, burner etc.. ) as key
+	//This will overwrite the value if the key is duplicated and return error if occurs
+	UpdateOneAddress(name, address string) error
```
Comment (favadi) [inline settings/address_storage.go]: is UpdateOneAddress too verbose? Maybe UpdateAddress is enough?

File: http/http_setting.go [hunk 25]
```diff
@@ -9,4 +12,6 @@ type Setting interface {
 	GetAllTokens() ([]common.Token, error)
 	NewTokenPair(base, quote string) (common.TokenPair, error)
 	UpdateToken(t common.Token) error
+	AddAddressToSet(setName string, address ethereum.Address) error
+	UpdateAdress(name string, address ethereum.Address) error
```
Comment (favadi) [inline http/http_setting.go]: typo :policeman:

UpdateAdress --> UpdateAddress.

File: settings/address_storage.go (hunk 24 above)
Comment (cyberliem) [reply settings/address_storage.go]: to distinguish from a future API, which might updateAllAddress at once
